### DISCUSSION TITLE
Mass upload block list CSV

### DISCUSSION THREAD

**Discussion by phillycodehound** [mastodon/mastodon#19672] [Q&A]
Is there a way for an Admin of an instance to mass upload a CSV of blocked servers to their Mastodon instance?

**Comment by enzocomics** [comment]
I couldn't find a way myself, and would also love for this to be possible, please. The current system of adding domains one by one is very time consuming!

Thanks.

Edit-- just found this in issues https://github.com/mastodon/mastodon/issues/18937#issue-1330694570

**Comment by tuetenk0pp** [comment]
I think this would somehow work with exporting/importing the domain_blocks table but unfortunately I'm no database expert so I can't supply the specific commands for that. All I can figure out is how to select the table in Postgres like this:

```sh
psql
\c mastodon_production
SELECT * FROM domain_blocks;
```

which gives:

```log
 id |    domain    |         created_at         |         updated_at         | severity | reject_media | reject_reports | private_comment | public_comment | obfuscate
----+--------------+----------------------------+----------------------------+----------+--------------+----------------+-----------------+----------------+-----------
  1 | catgirl.life | 2022-11-08 17:46:31.531791 | 2022-11-08 17:46:31.531791 |        1 | f            | f              |                 |                | f
```

I'm sure there is a way to export/import this between instances.

**Reply by tuetenk0pp** [reply]
OK I think I figured this out. It works with the COPY command (see [Documentation](https://www.postgresql.org/docs/15/sql-copy.html)). Here is my procedure:

**Please note: I am not sure if I would recommend this as there could database schema conflicts between Mastodon versions.**

1. log in to postgres and connect to the database with

```sh
\c mastodon_production
```

2.  Export the table ``domain_blocks`` with the COPY command:

```sh
COPY domain_blocks TO '/var/lib/postgresql/domain_blocks.csv' DELIMITER ';' CSV HEADER;
```

3. Go to the WebUI and delete the domain blocks
4. Verify that the corresponding table is empty:

```sh
SELECT * FROM domain_blocks;
```

5. Import the table with the COPY command:

```sh
COPY domain_blocks FROM '/var/lib/postgresql/domain_blocks.csv' DELIMITER ';' CSV HEADER;
```

6. Verify that the import worked by looking at the WebUI. Also check if the table now contains data with:

```sh
SELECT * FROM domain_blocks;
```

**Reply by tuetenk0pp** [reply]
@Gargron what do you think? Would you supply the domain_blocks table from mastodon.social? If yes/no, why?

**Comment by Deauthorized** [comment]
It'd be nice if this also had the capability to periodically auto-download from a remote server, like example.com/bl/nastyness.csv

Publicly sourced instance blocklists would do wonders for new spam instances popping up, without instance admins having to either find them via reports or whatever else.

Regex matching for domains to defed from would also be cool.

**Comment by cham2019** [comment]
I tried by installing Adminer. Then after entering manually like just a few of Blocked domains in the current database, I used Adminer interface to download the CSV file, and in excel I could edit that file adding almost 300 lines of bad server domains (I did my research, basically using many good instances block lists). Then I imported the new loaded CSV file into the table "domain_blocks". The change was not shown immediately. So I rebooted the server (you could try restarting just mastodon and PostgreSQL services). Then after a moment I saw the entire domain block list loaded in mastodon GUI. I hope that table be the only one where mastodon checks for the defederated instances. I hope it works.

**Comment by danakj** [comment]
Is this the answer to this question? https://fedi.tips/importing-ready-made-server-blocklists-on-mastodon/
